vendors/ti/boards/cc3220_launchpad/ports/common_io/iot_uart.c

```diff
@@ -0,0 +1,478 @@
+                              size_t bytes )
+{
+    if( ( uartHandle == NULL ) || ( pBuffer == NULL ) || ( bytes == 0 ) || ( uartHandle->tiUartHandle == NULL ) )
+    {
+        return IOT_UART_INVALID_VALUE;
+    }
+
+    int32_t ret = IOT_UART_SUCCESS;
+
+    if( UART_write( uartHandle->tiUartHandle, pBuffer, bytes ) < 0 )
+    {
+        ret = IOT_UART_WRITE_FAILED;
+    }
+
+    return ret;
+}
+
+/*-----------------------------------------------------------*/
+int32_t iot_uart_close( IotUARTHandle_t const uartHandle )
+{
+    if( ( uartHandle == NULL ) || ( uartHandle->tiUartHandle == NULL ) )
+    {
+        return IOT_UART_INVALID_VALUE;
+    }
+
+    /* TI Driver requires to cancel on-going transfer before calling close. */
+    iot_uart_cancel( uartHandle );
+
+    UART_close( uartHandle->tiUartHandle );
+    uartHandle->tiUartHandle = NULL;
+
+    return IOT_UART_SUCCESS;
+}
+
+/*-----------------------------------------------------------*/
+
+int32_t iot_uart_ioctl( IotUARTHandle_t const uartHandle,
+                        IotUARTIoctlRequest_t uartRequest,
+                        void * const pBuffer )
+{
+    if( ( uartHandle == NULL ) || ( pBuffer == NULL ) || ( uartHandle->tiUartHandle == NULL ) )
+    {
+        return IOT_UART_INVALID_VALUE;
+    }
+
+    int32_t ret = IOT_UART_SUCCESS;
+    const UARTCC32XX_HWAttrsV1 * hwAttrs = _getHWAttrs( uartHandle->tiUartHandle );
+
+    ClockP_FreqHz freq;
+    ClockP_getCpuFreq( &freq );
+
+    switch( uartRequest )
+    {
+        case eUartSetConfig:
+           {
+               if( _isBusy( uartHandle->tiUartHandle ) )
+               {
+                   ret = IOT_UART_BUSY;
+                   break;
+               }
+
+               IotUARTConfig_t * pConfig = ( IotUARTConfig_t * ) pBuffer;
+               unsigned long tiFlowMode = pConfig->ucFlowControl == 0 ? UART_FLOWCONTROL_NONE : UART_FLOWCONTROL_TX | UART_FLOWCONTROL_RX;
+               unsigned long tiConfig = dataLength[ pConfig->ucWordlength - 5 ] | stopBits[ pConfig->xStopbits ] | parityType[ pConfig->xParity ];
+
+               MAP_UARTFlowControlSet( hwAttrs->baseAddr, tiFlowMode );
+               MAP_UARTConfigSetExpClk( hwAttrs->baseAddr, freq.lo, pConfig->ulBaudrate, tiConfig );
+
+               break;
+           }
+
+        case eUartGetConfig:
+           {
+               IotUARTConfig_t * pConfig = ( IotUARTConfig_t * ) pBuffer;
+               unsigned long baud = 0;
+               unsigned long tiConfig = 0;
+               unsigned long tiFlowMode = 0;
+
+               tiFlowMode = MAP_UARTFlowControlGet( hwAttrs->baseAddr );
+               MAP_UARTConfigGetExpClk( hwAttrs->baseAddr, freq.lo, &baud, &tiConfig );
+
+               pConfig->ucFlowControl = tiFlowMode == UART_FLOWCONTROL_NONE ? 0 : 1;
+               pConfig->ulBaudrate = baud;
+               pConfig->ucWordlength = ( uint8_t ) _indexOf( dataLength, sizeof( dataLength ), ( tiConfig & UART_CONFIG_WLEN_MASK ) ) + 5;
```

Comment:
Note that `_indexOf()` returns `int8_t` and can return `-1`.  If that happens, `pConfig->ucWordLength` will be set to `0x4` here (cast to `uint8_t` gives `0xff`, then + 5).  Is that what we want?